src/main/scala/cognite/spark/v1/AssetHierarchyBuilder.scala

```diff
@@ -174,86 +124,55 @@ class AssetHierarchyBuilder(config: RelationConfig)(val sqlContext: SQLContext)
     } yield ()
   }
 
-  def validateSubtreeRoots(roots: Seq[AssetsIngestSchema]): IO[Unit] = {
+  def validateSubtreeRoots(roots: Vector[AssetsIngestSchema]): IO[Unit] =
     // check that all `parentExternalId`s exist
-    val ids = roots.map(_.parentExternalId).filter(_.nonEmpty)
-    if (ids.isEmpty) {
-      IO.unit
-    } else {
+    batchedOperation[String, Asset](
+      roots.map(_.parentExternalId).filter(_.nonEmpty).distinct,
       // The API calls throw exception when any of the ids do not exist
-      ids.distinct
-        .grouped(batchSize)
-        .toVector
-        .parTraverse { idsInGroup =>
-          client.assets
-            .retrieveByExternalIds(idsInGroup)
-            .adaptError({
-              case e: CdpApiException if e.code == 400 && e.missing.nonEmpty =>
-                val missingNodes = e.missing.get.map(j => j("externalId").get.asString.get).take(10)
-                val referencingNodes =
-                  missingNodes
-                  // always take the one with "lowest" externalId, so the errors are deterministic
-                    .map(missing => roots.filter(_.parentExternalId == missing).minBy(_.externalId))
-                    .map(_.externalId)
-                InvalidNodeReferenceException(missingNodes, referencingNodes)
```

Comment:
We now checking for this error before we even do the insert, so no need for this complex error-parsing logic :)